pkg/client/client.go

```diff
@@ -751,33 +691,34 @@ func (d *DynatraceClient) DeleteSettings(objectID string) error {
 
 const emptyResponseRetryMax = 10
 
-func isRetryOnEmptyResponse(receivedCount int, emptyResponseRetryCount *int, resp rest.Response) (bool, error) {
+func isRetryOnEmptyResponse(receivedCount int, emptyResponseRetryCount int, resp rest.Response) (bool, int, error) {
 	if receivedCount == 0 {
-		if *emptyResponseRetryCount < emptyResponseRetryMax {
-			*emptyResponseRetryCount++
-			throttleTooManyRequests("Received empty array response, retrying with same nextPageKey (HTTP: %d) ", resp.StatusCode)
-			return true, nil
+		if emptyResponseRetryCount < emptyResponseRetryMax {
+			emptyResponseRetryCount++
+			rateLimitStrategy := rest.CreateRateLimitStrategy()
+			rateLimitStrategy.ThrottleCallAfterError("Received empty array response, retrying with same nextPageKey (HTTP: %d) ", resp.StatusCode)
```

Comment:
hmm.. I actually meant that we could possibly move the logic of these methods into a new implementation of RateLimitStrategy that is then used for the entity download.

If that is not possible (didn't check how the code would look, where the strategy is created right now), I would actually suggest you stick to just the sleep as part of the client. 
I would then move this code into entities_client at least though, as it's entity specific and client.go is already huge

Reply:
I'll see what can be done

Reply:
BTW, it is not Entities Specific, well I don't believe so
I'm pretty sure it could happen for Settings as well, given an big enough load on the cluster

Reply:
If it did happen without my code present, you couldn't notice, as an empty response would not trigger errors

Reply:
I extracted the shared throttling code from rateLimiter to an internal source
This will safeguard the integrity of the rateLimiter's original purpose
While allowing me to throttle calls made after specific errors

If I don't do this, then I'm creating unnecessary 403s, which will force users to lower their concurrent requests, and restart their lengthy extractions after waiting for 5-10minutes

The real problem is that the rateLimiter only acts after we have busted the limit, it doesn't limit the rate at all.
A real rateLimiter would limit the global amount of calls 'per minute' for all threads, making that we never bust the limit, or at least try not to as we are not the only users of the clusters

Then I wouldn't need to add individual throttling for all of my error handling
And we could use as many ConcurrentRequests as we want because they would be limited anyways